Encode codemap and span information in crate metadata.

michaelwoerister opened the pull request:
This allows to create proper debuginfo line information for items inlined from other crates (e.g. instantiations of generics). Only the codemap's 'metadata' is stored in a crate's metadata. That is, just filename, positions of line-beginnings, etc. but not the actual source code itself.

Crate metadata size is increased by this change because spans in the encoded ASTs take up space now:
```
                BEFORE    AFTER
libcore         36 MiB    39.6 MiB    +10%
libsyntax       51.1 MiB  60.5 MiB    +18.4%
libcollections  11.2 MiB  12.8 MiB    +14.3%
```
This only affects binaries containing metadata (rlibs and dylibs), executables should not be affected in size. 

Fixes #19228 and probably #22226.

steveklabnik commented:
r?

alexcrichton commented:
Do you know if it would be easy enough to track which spans are associated with generic functions and which aren't? It'd be nice to just cut down on the metadata size slightly (4mb for libcore is pretty big)

michaelwoerister commented:
I'll see what I can do to omit span data for types and other stuff that is not used later anyway. I'll also experiment with encoding spans in LEB128 and see if that helps in addition to the LZ77 encoding we are doing anyway.

I'll be on vacation for the rest of the week, however, so please don't expect any updates before next week.

Manishearth commented:
Needs a rebase, sorry.

michaelwoerister commented:
So, I've been investigating different options of reducing space usage for spans but I've not been really successful.
+ At the moment it's not possible to omit some spans from metadata but not others. The reason is that spans occur in the AST and we use `derive`-generated `Encode/Decode` implementations for these data structures. If we had a `#[dont_serialize_me]` attributes for fields, we could do something here. However, I didn't want to add this feature in a rudimentary, ad-hoc fashion, just for this one use case.
+ I made the compiler store spans in the variable-sized LEB128 format also used in DWARF. This however didn't give the expected space reduction because our RBML writer is super-wasteful, tagging each `u8` I emit with a type-tag and a byte storing the size of the number. This results in at least 3 bytes and up to 12 bytes per `BytePos` and between 6 and 24 bytes per `Span` which totally defeats the purpose of storing things in LEB128.

A maybe acceptable intermediate solution is to store spans as one `u64` in order to cut down on RBML tagging overhead. In a preliminary test this worked quite well under the circumstances. I'll push a cleaned up version of this the day after tomorrow.

The real problem though is how metadata is encoded in general (an issue already recorded in #21482). It's super-verbose and I doubt that all this tagging really has much of a benefit. Coming up with a better data format will be necessary sooner or later.

alexcrichton commented:
Wow that is... depressing. It's not necessarily the end of the world to inflate metadata size right now, it's just something to think about :). If you've got a smaller format working though, that sounds great!

michaelwoerister commented:
> Wow that is... depressing.

That's one way to see it. I see the current metadata encoding as a lustrous garden of low-hanging fruit, overripe and ready for the picking. How often does one get the chance to shave off a third of the space consumption at no cost at all.

We'll have to pin down the requirements for the metadata format though before the feast can start.

michaelwoerister commented:
OK, this is updated now. Numbers are better than before. 
```
                BEFORE    AFTER
libcore         39   MiB  41   MiB    +5%
libsyntax       54.6 MiB  61.2 MiB    +12%
libcollections  14.3 MiB  15.3 MiB    +7%
```

alexcrichton commented:
Ok, this is all looking good to me, nice work @michaelwoerister!

I'd like a second set of eyes on this, however.

r? @nrc or @brson

nrc commented:
I need to give this a proper read, but I think the idea is good.

One question: what about macro expansions? I guess we don't use them in debug info? Is it ever possible we'd want to have this information cross-crate?

michaelwoerister commented:
Yes, macro expansion information is not used in debuginfo. DWARF has some provisions for recording C-macro definitions (not expansions) but that's not even supported by LLVM. I guess this won't be a topic for us in the foreseeable future.

About having this information cross-crate in general: It would be easy to implement, I think, but it costs additional space and at the moment I don't know of a use case. For now, I'd just leave it out.

nrc commented on src/librustc/metadata/creader.rs:
```diff
@@ -586,3 +588,131 @@ impl<'a> CrateReader<'a> {
+            }
+        }
+
+        return true;
```
nit: unnecessary `return`

nrc commented on src/librustc/metadata/decoder.rs:
```diff
@@ -1564,3 +1563,19 @@ pub fn is_default_trait<'tcx>(cdata: Cmd, id: ast::NodeId) -> bool {
     }
 }
+
+pub fn get_codemap(metadata: &[u8]) -> Vec<codemap::FileMap> {
```
this could have a better name since it doesn't actually return a codemap

nrc commented on src/librustc/middle/astencode.rs:
```diff
@@ -233,8 +243,50 @@ impl<'a, 'b, 'tcx> DecodeContext<'a, 'b, 'tcx> {
+    /// within the local crate's codemap. `creader::import_codemap()` will
+    /// already have allocated any additionally needed FileMaps in the local
+    /// codemap.
+    pub fn tr_span(&self, span: Span) -> Span {
```
it would be good to s/tr_span/translate_span. If this is a huge amount of work, don't worry about it (I see it was called this before).

nrc commented on src/librustc/middle/astencode.rs:
```diff
@@ -233,8 +243,50 @@ impl<'a, 'b, 'tcx> DecodeContext<'a, 'b, 'tcx> {
-        codemap::DUMMY_SP // FIXME (#1972): handle span properly
+
+    /// Translates a `Span` from an extern crate to the corresponding `Span`
+    /// within the local crate's codemap. `creader::import_codemap()` will
```
this 'will' is a bit worrying to me - who's responsibility is it to ensure this invariant? Is it possible to cause an error by using this function without establishing that invariant.

nrc replied:
I assume this is just a nit about the comment, I don't expect you'll need to change any code, just clarify the comment

nrc commented on src/librustc/middle/astencode.rs:
```diff
@@ -144,7 +151,10 @@ pub fn decode_inlined_item<'tcx>(cdata: &cstore::crate_metadata,
             from_id_range: from_id_range,
-            to_id_range: to_id_range
+            to_id_range: to_id_range,
+            translate_spans: cdata.codemap_import_info.len() > 0 &&
```
would it be harmful to always translate spans? I don't understand why the first predicate would be false - is this for backwards compatibility with existing crates? If we do the translate work and it is unnecessary (second predicate) - is that a real waste of resources? Does if actually introduce an error?

In general, I like to avoid flags like this - I feel the increased scope for error outweighs the performance benefit. But if it is necessary, then leave it.

nrc commented on src/libsyntax/codemap.rs:
nice :-)

nrc commented on src/libsyntax/codemap.rs:
Nit^2: Comments should be proper sentences - start with a capital, end with a full stop (sorry for nittiness).

nrc commented:
r+ with the nits addressed

michaelwoerister commented:
Thanks for the review. I work in your comments on Wednesday. Don't worry about nittiness `:)`

lifthrasiir commented:
The space optimization for filemap may have to be revisited after #22971 lands. The relevant commit is fe73d382eeabaed0c37425388ab15ad690e5906f, which makes the tag represent the size of following integers. If you keep individual tags in the filemap then you don't need separate `bytes_per_diff`; otherwise you will need `Opaque` to elide tags at all.

bors commented:
:umbrella: The latest upstream changes (presumably #22971) made this pull request unmergeable. Please resolve the merge conflicts.

michaelwoerister commented:
I think I've addressed all comments. The encoding of the line data in filemaps is not optimal yet. I started implementing a better version using `emit_opaque()` as @lifthrasiir suggested but it got so ugly and unwieldy that I decided to do this in a separate PR later.

michaelwoerister commented:
@bors r=nrc

bors commented:
@bors r=michaelwoerister 2f88655

bors commented:
:hourglass: Testing commit 2f88655 with merge 3b3bb0e...

bors commented:
:sunny: Test successful - [auto-linux-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-32-nopt-t/builds/3936), [auto-linux-32-opt](http://buildbot.rust-lang.org/builders/auto-linux-32-opt/builds/3941), [auto-linux-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-linux-64-nopt-t/builds/3929), [auto-linux-64-opt](http://buildbot.rust-lang.org/builders/auto-linux-64-opt/builds/3933), [auto-linux-64-x-android-t](http://buildbot.rust-lang.org/builders/auto-linux-64-x-android-t/builds/3903), [auto-mac-32-opt](http://buildbot.rust-lang.org/builders/auto-mac-32-opt/builds/3957), [auto-mac-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-mac-64-nopt-t/builds/3958), [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3945), [auto-win-32-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-32-nopt-t/builds/3613), [auto-win-32-opt](http://buildbot.rust-lang.org/builders/auto-win-32-opt/builds/3603), [auto-win-64-nopt-t](http://buildbot.rust-lang.org/builders/auto-win-64-nopt-t/builds/2432), [auto-win-64-opt](http://buildbot.rust-lang.org/builders/auto-win-64-opt/builds/2439)
